### PR TITLE
[ConvDiff][Fluid] Buoyancy and conjugate heat transfer solvers

### PR DESCRIPTION
**Pull request by rubenzorrilla**
@RiccardoRossi  here it is. @jcotela , have a look to it as well (some changes belong to the FluidDynamicsApplication).

### CODE REVIEW

**Comment by RiccardoRossi** [comment]
is this fixed?

**Comment by rubenzorrilla** [comment]
@RiccardoRossi  what?

**Review by RiccardoRossi — COMMENTED** [review]
can anyone in @KratosMultiphysics/technical-committee  agree about the modification in modelpart.h?

**Comment by RiccardoRossi** [comment]
@rubenzorrilla travis does not compile...

File: kratos/includes/model_part.h [hunk 1]
```diff
@@ -515,6 +515,11 @@ class KRATOS_API(KRATOS_CORE) ModelPart : public DataValueContainer, public Flag
                 << "\" to the model part with name \"" << this->Name() << "\" which is not empty" << std::endl;
 
             mpVariablesList->Add(ThisVariable);
+
+            for(auto it = this->NodesBegin(); it!= this->NodesEnd(); ++it)
```
Comment (philbucher) [inline kratos/includes/model_part.h]: why is this needed here?
Isn't this handled somewhere else?

File: kratos/python_scripts/python_solver.py [hunk 2]
```diff
@@ -157,6 +157,8 @@ def _ImportModelPart(self, model_part, model_part_import_settings):
             from restart_utility import RestartUtility
             RestartUtility(model_part, self._GetRestartSettings(model_part_import_settings)).LoadRestart()
             self.print_on_rank_zero("::[PythonSolver]::", "Finished loading model part from restart file.")
+        elif(input_type == "use_input_model_part"):
```
Comment (philbucher) [inline kratos/python_scripts/python_solver.py:160]: why is this needed?
In such cases I would not call `ImportModelPart`

I suggest doing sth like we do in [StructuralMechanics](https://github.com/KratosMultiphysics/Kratos/blob/master/applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py#L204-L208) to not do modifications in this class

File: kratos/containers/model.cpp [hunk 3]
```diff
@@ -72,7 +72,7 @@ namespace Kratos
         else
         {
             if(&(*search->second) != &*(pModelPart))
-                KRATOS_ERROR << "trying to add to the Model two DISTINCT model parts with the same name. This should be possible (and it will be in the future) if they belong to two different root model_parts, but it is currently disallowed";
+                KRATOS_WARNING("Model") << "trying to add to the Model two DISTINCT model parts with the same name. This should be possible (and it will be in the future) if they belong to two different root model_parts, but it is currently disallowed" << std::endl;
```
Comment (philbucher) [inline kratos/containers/model.cpp]: :+1:

**Review by philbucher — COMMENTED** [review]
@rubenzorrilla this PR silently adds / changes some important things in the core
I am not blocking but please make separate PRs in the future for such things

File: kratos/utilities/merge_variable_lists_utility.h [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+  ///@{
+      
+  ///@}
+  ///@name Kratos Classes
+  ///@{
+  
+  /// Short class definition.
+  /** Detail class definition.
+  */
+  class MergeVariableListsUtility
```
Comment (philbucher) [inline kratos/utilities/merge_variable_lists_utility.h]: - please add some documentation
- and maybe a test
- splitting into header&source?
- there are leftover debugging statements

File: kratos/sources/connectivity_preserve_modeler.cpp [hunk 5]
```diff
@@ -53,9 +55,37 @@ void ConnectivityPreserveModeler::GenerateModelPart(
 }
 
 // Private methods /////////////////////////////////////////////////////////////
+void ConnectivityPreserveModeler::CheckVariableLists(ModelPart &rOriginModelPart, ModelPart &rDestinationModelPart)
```
Comment (philbucher) [inline kratos/sources/connectivity_preserve_modeler.cpp:58]: leftover debugging in this fct

File: kratos/sources/connectivity_preserve_modeler.cpp [hunk 6]
```diff
@@ -53,9 +55,37 @@ void ConnectivityPreserveModeler::GenerateModelPart(
...
+        {
+          std::cout << var << std::endl;
+        }
+
+    for(const auto& var : rdestination_variable_list)
+        if(rorigin_variable_list.Has(var) == false)
+            KRATOS_ERROR << "Variable: " << var << " is in rDestinationModelPart variables but not in the rOriginModelPart variables" << std::endl;
+
+    for(const auto& var : rorigin_variable_list)
+        if(rdestination_variable_list.Has(var) == false)
```
Comment (philbucher) [inline kratos/sources/connectivity_preserve_modeler.cpp]: you could use `KRATOS_ERROR_IF_NOT(rdestination_variable_list.Has(var)) `

**Review by philbucher — COMMENTED** [review]
some more cleanup necessary

File: kratos/includes/model_part.h (hunk 1 above)
Comment (roigcarlo) [reply kratos/includes/model_part.h]: Agree with that, please wait until we discuss this in more detail.

**Review by roigcarlo — DISMISSED** [review]
See the comment about the modelpart

File: kratos/python_scripts/python_solver.py (hunk 2 above)
Comment (rubenzorrilla) [reply kratos/python_scripts/python_solver.py:160]: In both the Buoyancy solver and the CHT solver, the thermal fluid model part is set from the CFD solver one. Then, the import is skipped in this cases. However, if you are solving a simple heat transfer problem, you need to import the model part as usual. To avoid reimplementing the entire ImportModelPart() method in the convection-diffusion solver including this if statement, we decided to add this option. IMO it does not harm and might be useful for similar developments.

Comment (jcotela) [reply kratos/python_scripts/python_solver.py:160]: @philbucher if we decide to go with what you are doing in the StructuralMechanics, I'd suggest moving that (the member variable and the auxiliary function) to the base class.

File: kratos/utilities/merge_variable_lists_utility.h (hunk 4 above)
Comment (rubenzorrilla) [reply kratos/utilities/merge_variable_lists_utility.h]: Sure! (This was done in a rush by @RiccardoRossi to make things work)

**Comment by rubenzorrilla** [comment]
@philbucher If you have a look to my changes, they are only related to the CHT solver, which was the aim of this branch. You know that I do not like such large PR neither...

@RiccardoRossi I can attend @philbucher 's demands.

@roigcarlo @philbucher concerning de model_part.h changes, I also think that we need further discussion but let's try to not delay this too much since we need this solver for an industrial project.

**Comment by rubenzorrilla** [comment]
And thank you all for the review!

**Comment by philbucher** [comment]
> @roigcarlo @philbucher concerning de model_part.h changes, I also think that we need further discussion but let's try to not delay this too much since we need this solver for an industrial project.

if you are in a hurry why don't you create your own branch and afterwards slowly merge the changes to master?

File: kratos/includes/model_part.h [hunk 7]
```diff
@@ -515,6 +515,11 @@ class KRATOS_API(KRATOS_CORE) ModelPart : public DataValueContainer, public Flag
                 << "\" to the model part with name \"" << this->Name() << "\" which is not empty" << std::endl;
 
             mpVariablesList->Add(ThisVariable);
+
```
Comment (philbucher) [inline kratos/includes/model_part.h]: Actually this should not be done
It even throws if nodes are present, see the lines above
We had a lot of memory problems, see #492

**Review by philbucher — DISMISSED** [review]
See comment, this change might cause memory issues!

**Comment by pooyan-dadvand** [comment]
The change in model part is very tricky and only works in a very controlled conditions. Let say, in most of the cases is not a good idea to change the variable lists of other model parts.

File: kratos/python_scripts/python_solver.py (hunk 2 above)
Comment (philbucher) [reply kratos/python_scripts/python_solver.py:160]: Yep whatever will be decided here I will also adapt in StructuralMechanics (FYI @dbaumgaertner @armingeiser )

File: kratos/sources/connectivity_preserve_modeler.cpp (hunk 6 above)
Comment (RiccardoRossi) [reply kratos/sources/connectivity_preserve_modeler.cpp]: about this ... even if IT IS possible to use KRATOS_ERROR_IF_NOT i personally don't see any problem with not using that... 

I mean...it is a question of taste, but in my opinion no harm it is done by putting the if explicitly

Comment (philbucher) [reply kratos/sources/connectivity_preserve_modeler.cpp]: agree, thats why I wrote "could use"

File: kratos/python_scripts/python_solver.py (hunk 2 above)
Comment (RiccardoRossi) [reply kratos/python_scripts/python_solver.py:160]: the original idea (documented for example here https://github.com/KratosMultiphysics/Kratos/wiki/Example:-A-simple-Fluid---Thermal-coupled-solver)

was that on import one can have one of 

* "input_type": "mdpa",
* "input_type": "use_input_model_part",
* "input_type": "Restart"

if as here one uses "use_input_model_part" then an existing modelpart is read and cloned.

I am very open to changing this ... let's just agree on a standardized form which provides equivalent feature and document it

File: kratos/utilities/merge_variable_lists_utility.h (hunk 4 above)
Comment (RiccardoRossi) [reply kratos/utilities/merge_variable_lists_utility.h]: just to tell that this "merges" the variable lists of different modelparts.

Essentially it solves in a "good" way the "hack" described in (documented for example here https://github.com/KratosMultiphysics/Kratos/wiki/Example:-A-simple-Fluid---Thermal-coupled-solver)

just to document what the problem is imagine that you have a mp1 and a mp2 initially empty but which will be eventually filled with the same nodes as mp1

one could do something like

    solver1(mp1).AddVariables()
    solver2(mp2).AddVariables()

the solver2 only addes variables to mp2, NOT TO MP1, so when the nodes are created (by mp1) they simply don't have the variables added by mp2

File: kratos/includes/model_part.h (hunk 1 above)
Comment (RiccardoRossi) [reply kratos/includes/model_part.h]: I tend to agree about this. Calling Merge on the variable lists after AddVariables is a better solution. However if we add a variable, and the variable is not in the node list it is a potentially very big problem

**Review by RiccardoRossi — COMMENTED** [review]
changes in model_part.h were removed - this should fix the most conflictive problem in this PR

File: kratos/containers/model.cpp (hunk 3 above)
Comment (philbucher) [reply kratos/containers/model.cpp]: @RiccardoRossi after thinking abt this some more, what happens if a modelpart with the same name is used, will the existing one be overwritten?

Comment (RiccardoRossi) [reply kratos/containers/model.cpp]: the answer is ...NOTHING ...meaning that if a modelpart with the same name is already in model nothing new is added, that is, the old one remains untouched and the new one is not added

File: kratos/python_scripts/python_solver.py (hunk 2 above)
Comment (RiccardoRossi) [reply kratos/python_scripts/python_solver.py:160]: @philbucher i am not clear of what is your proposal. could u explain it a bit here?

File: kratos/utilities/merge_variable_lists_utility.h (hunk 4 above)
Comment (RiccardoRossi) [reply kratos/utilities/merge_variable_lists_utility.h]: documentation was added as well as a test. the leftover debugging output was removed.

with regards to splitting in header and source i don't really see the need of it...it is a rather small function... do you see it really important?

File: kratos/containers/model.cpp (hunk 3 above)
Comment (RiccardoRossi) [reply kratos/containers/model.cpp]: additionally to what i am telling here, the function AddModelPart should eventually be removed once we ultimatelly upgrade the model (which i still hope will happen soon)

File: kratos/utilities/merge_variable_lists_utility.h (hunk 4 above)
Comment (loumalouomega) [reply kratos/utilities/merge_variable_lists_utility.h]: I agree with @philbucher to split into header/source. to get the benefits of using cotire code must be separated in header/source

File: kratos/containers/model.cpp (hunk 3 above)
Comment (armingeiser) [reply kratos/containers/model.cpp]: Is there a special reason this can't stay an error until then?

File: kratos/python_scripts/python_solver.py (hunk 2 above)
Comment (armingeiser) [reply kratos/python_scripts/python_solver.py:160]: FYI @MFusseder

File: kratos/containers/model.cpp (hunk 3 above)
Comment (RiccardoRossi) [reply kratos/containers/model.cpp]: in any case i just brought this back to the master version...

File: kratos/sources/connectivity_preserve_modeler.cpp (hunk 5 above)
Comment (RiccardoRossi) [reply kratos/sources/connectivity_preserve_modeler.cpp:58]: cleaned up

**Review by RiccardoRossi — COMMENTED** [review]
I believe all the conflictive changes were removed and this can go in ... 

the only leftover is the discussion is about how to do the import (@philbucher i name you here), which deserves a discussion and a separate PR.

@roigcarlo would u be willing to remove your blocker?

File: kratos/python_scripts/python_solver.py (hunk 2 above)
Comment (philbucher) [reply kratos/python_scripts/python_solver.py:160]: @RiccardoRossi I think I like your solution more, so discard my comment
having a special `input_type` that does nothing (i.e. not import the ModelPart from a file) seems like a good solution on second thought
personally I think the name `use_existing_model_part` is more clear, but `use_input_model_part` would also be ok

once this is merged I will adapt StructuralMechanics to what we agree on here
@armingeiser @dbaumgaertner @MFusseder this means that you will have to do a minor modification in the ShapeOptApp. I will let you know

@RiccardoRossi To sum up, I am fine with the changes in this PR, from my side no separate PR is necessary

**Review by philbucher — COMMENTED** [review]
no further comments from my side

I am not approving bcs I am not familiar with the ConvDiffApp

**Comment by philbucher** [comment]
@RiccardoRossi there are some compiler warnings, thats why travis fails

File: applications/FluidDynamicsApplication/custom_processes/boussinesq_force_process.cpp [hunk 8]
```diff
@@ -96,34 +96,32 @@ namespace Kratos
...
+        KRATOS_ERROR_IF_NOT( mrModelPart.GetProcessInfo().Has(AMBIENT_TEMPERATURE) ) <<
         "In Boussinesq Force Process: \'AMBIENT_TEMPERATURE\' not given in ProcessInfo." << std::endl;
     }
 
     void BoussinesqForceProcess::AssignBoussinesqForce()
     {
-        ModelPart &rModelPart = *mpModelPart;
-
-        const double AmbientTemperature = rModelPart.GetProcessInfo().GetValue(AMBIENT_TEMPERATURE);
+        const double AmbientTemperature = mrModelPart.GetProcessInfo().GetValue(AMBIENT_TEMPERATURE);
```
Comment (loumalouomega) [inline applications/FluidDynamicsApplication/custom_processes/boussinesq_force_process.cpp]: snake _case?

File: applications/FluidDynamicsApplication/custom_processes/boussinesq_force_process.cpp [hunk 9]
```diff
@@ -96,34 +96,32 @@ namespace Kratos
...
         KRATOS_ERROR_IF( AmbientTemperature <= 0.0 ) <<
         "In Boussinesq Force Process: \'AMBIENT_TEMPERATURE\' obtained from ProcessInfo is incorrect." << std::endl <<
         "Expected a positive double, got " << AmbientTemperature << std::endl;
 
         // Note: the default value of 1/AMBIENT_TEMPERATURE is the usual assumption for perfect gases.
         const double Alpha = (mUseAmbientTemperature) ? 1.0 / AmbientTemperature : mThermalExpansionCoefficient;
 
-        int NumNodes = rModelPart.NumberOfNodes();
+        int NumNodes = mrModelPart.NumberOfNodes();
```
Comment (loumalouomega) [inline applications/FluidDynamicsApplication/custom_processes/boussinesq_force_process.cpp]: snake_case?

File: applications/convection_diffusion_application/convection_diffusion_application_variables.cpp [hunk 10]
```diff
@@ -13,17 +13,19 @@
...
 {
-KRATOS_CREATE_VARIABLE(double, MELT_TEMPERATURE_1)
-KRATOS_CREATE_VARIABLE(double, MELT_TEMPERATURE_2)
+
+KRATOS_CREATE_VARIABLE(double, AUX_FLUX)
+KRATOS_CREATE_VARIABLE(double, AUX_TEMPERATURE)
 KRATOS_CREATE_VARIABLE(double, BFECC_ERROR)
 KRATOS_CREATE_VARIABLE(double, BFECC_ERROR_1)
-
+KRATOS_CREATE_VARIABLE(double, DELTA_SCALAR1)
```
Comment (loumalouomega) [inline applications/convection_diffusion_application/convection_diffusion_application_variables.cpp:21]: The names are not very descriptive (why 1 and 2)?

File: applications/convection_diffusion_application/python_scripts/conjugate_heat_transfer_solver.py [hunk 11]
```diff
@@ -0,0 +1,305 @@
...
+# Check that applications were imported in the main script
+KratosMultiphysics.CheckRegisteredApplications("FluidDynamicsApplication")
+KratosMultiphysics.CheckRegisteredApplications("ConvectionDiffusionApplication")
+
+# Import applications
+import KratosMultiphysics.FluidDynamicsApplication as KratosCFD
+import KratosMultiphysics.ConvectionDiffusionApplication as KratosConvDiff
+# TODO: USE THE MAPPING APPLICATION INSTEAD OF THE FSI MAPPER
+import KratosMultiphysics.FSIApplication as FSIApplication
+import NonConformant_OneSideMap as ncosm
```
Comment (loumalouomega) [inline applications/convection_diffusion_application/python_scripts/conjugate_heat_transfer_solver.py:16]: Why not using mortar mapper?

**Review by loumalouomega — COMMENTED** [review]
Minor comments

File: applications/convection_diffusion_application/python_scripts/conjugate_heat_transfer_solver.py (hunk 11 above)
Comment (RiccardoRossi) [reply applications/convection_diffusion_application/python_scripts/conjugate_heat_transfer_solver.py:16]: We can discuss about that, however this is best done in a separate PR. Why do we have both Mortar Mapper and the NonConformant whatever map? do they at least have the same interface?

i would open an issue to discuss about that

Comment (loumalouomega) [reply applications/convection_diffusion_application/python_scripts/conjugate_heat_transfer_solver.py:16]: As you see I am not blocking, just asking. The interface is similar, but not the same. The mortar mappper converges always in one iteration (maximum two), because solves exactly the problem

Comment (philbucher) [reply applications/convection_diffusion_application/python_scripts/conjugate_heat_transfer_solver.py:16]: I believe there is even an application for this ...

Does not yet have the fancy mappers yet though

Comment (RiccardoRossi) [reply applications/convection_diffusion_application/python_scripts/conjugate_heat_transfer_solver.py:16]: as i am telling, you have a point. open an issue and we can discuss that in general...not as part of this PR in any case

Comment (loumalouomega) [reply applications/convection_diffusion_application/python_scripts/conjugate_heat_transfer_solver.py:16]: Ok

File: applications/FluidDynamicsApplication/custom_processes/boussinesq_force_process.cpp (hunk 9 above)
Comment (RiccardoRossi) [reply applications/FluidDynamicsApplication/custom_processes/boussinesq_force_process.cpp]: done (all of the function)

**Review by loumalouomega — DISMISSED** [review]
OK, then

File: applications/convection_diffusion_application/convection_diffusion_application_variables.cpp (hunk 10 above)
Comment (rubenzorrilla) [reply applications/convection_diffusion_application/convection_diffusion_application_variables.cpp:21]: These variables were already there before this PR implementations. They can be renamed in another PR if needed.

Comment (loumalouomega) [reply applications/convection_diffusion_application/convection_diffusion_application_variables.cpp:21]: Yes I was just asking

**Review by loumalouomega — APPROVED** [review]
OK, thanks for taking into account my comments
